### DISCUSSION TITLE
[API Proposal]: Switch expression without parent object

### DISCUSSION THREAD

**Discussion by simon-curtis** [dotnet/csharplang#5876] [Language Ideas]
### Background and motivation

I really enjoy switch expressions and I have found myself refactoring alot of if/else's to use them. 

I have found that when refactoring the logic doesnt tend to revolve around one single variable, this means I end up using a the switch statement with a random variable and doing `_ when ... =>` the whole way down the chain. i.e.

```C#
var meterId = message switch
{
    _ when message.TryFindChild("OldMeter", out var om) && om.TryFindChild(METER_ID, out var omid) => omid.InnerText,
    _ => RawString.Empty
}
```

I think it would be nice to able to have a switch expression that is essentially a list of expressions that are evaluated in sequence and escapes early. Maybe in this case it should under a different name to avoid confusion with current switch statments. Maybe cascade or evaluate (/eval)?

Of course right now you could foreach a list of conditions and break on the first one but thats a lot of syntax - and I think you'd agree the switch expression tree is nicer.

### API Proposal

N/A

### API Usage

```C#
var message = new U8Xml.XmlNode();

var output = eval {
   message.TryFindChild(METER_ID, out var meterIdNode) => meterIdNode.InnerText,
   _ => RawString.Empty
};
```

**Comment by hopperpl** [comment]
I think this would work

``` C#
var message = new U8Xml.XmlNode();

var output = true switch {
   true when message.TryFindChild(METER_ID, out var meterIdNode) => meterIdNode.InnerText,
   _ => RawString.Empty
};
```

the "true when" addition bypasses the constant expression constraint. So if you want that, you can already do it. For me this looks a lot more cryptic than an if-then-else chain, or using the ternary operator.

**Reply by Xyncgas** [reply]
ohh

**Comment by simon-curtis** [comment]
@hopperpl the example I have given is a trivial one for easy of reading. In a recent refactor I had multiple if/else if/else branches that once put into switch statements took up a total of 25 lines instead of ~90 and was far easier to reason then to keep remembering what if statement you were in and why.

**Reply by Xyncgas** [reply]
Switch also provides performance benefits sometimes

**Reply by HaloFour** [reply]
> Switch also provides performance benefits sometimes

Really only if you switch over a specific value and the compiler can convert that into a jump table.  That doesn't help for a serious of cascading comparisons, the compiler would effectively be emitting the same IL as if you had written a bunch of `if`/`else` statements.

**Comment by HaloFour** [comment]
I know this has come up before but it's probably buried in the comments of the various other `switch` expression discussions.  I recall that the team's position was that this is a better use case for the ternary conditional operator:

```cs
var output = message.TryFindChild(METER_ID, out var meterIdNode) ? meterIdNode.InnerText :
        SomeOtherCondition(METER_ID, out var otherMeterIdNode) ? otherMeterIdNode.InnerText :
        RawString.Empty;
```

**Reply by simon-curtis** [reply]
Thanks, but again this was a simple example that I gave. Turnary operators start getting very messy once you get above 2 paths

**Reply by HaloFour** [reply]
That is why I added a third, adding whitespace goes a long way to make nested conditional operations quite readable.
